Helper/Discount.php

```diff
@@ -0,0 +1,320 @@
+            // ignore the exception and proceed
+        }
+    }
+
+    /**
+     * Try to clear Amasty Gift Cart data for the unused immutable quotes and the parent quote
+     *
+     * @param int|string $immutableQuoteId
+     * @param int|string $parentQuoteId
+     */
+    public function deleteRedundantAmastyGiftCards($immutableQuoteId, $parentQuoteId)
+    {
+        try {
+            $connection = $this->resource->getConnection();
+            $giftCardTable = $this->resource->getTableName('amasty_amgiftcard_quote');
+            $quoteTable = $this->resource->getTableName('quote');
+
+            $sql = "DELETE FROM {$giftCardTable} WHERE quote_id IN 
+                    (SELECT entity_id FROM {$quoteTable} 
+                    WHERE bolt_parent_quote_id = :bolt_parent_quote_id AND entity_id != :entity_id)";
+
+            $connection->query($sql, ['entity_id' => $immutableQuoteId, 'bolt_parent_quote_id' => $parentQuoteId]);
+        } catch (\Zend_Db_Statement_Exception $e) {
+            // The amasty_amgiftcard_quote table is missing, Amasty Gift Cart is not installed,
+            // ignore the exception and proceed
+        }
+    }
+
+    /**
+     * Remove Amasty Gift Card and update quote totals
+     *
+     * @param int $codeId
+     * @param Quote $quote
+     */
+    public function removeAmastyGiftCard($codeId, $quote)
+    {
+        $connection = $this->resource->getConnection();
+        $giftCardTable = $this->resource->getTableName('amasty_amgiftcard_quote');
+
+        $sql = "DELETE FROM {$giftCardTable} WHERE code_id = :code_id AND quote_id = :quote_id";
+        $connection->query($sql, ['code_id' => $codeId, 'quote_id' => $quote->getId()]);
+
+        $this->updateTotals($quote);
+    }
+
+    /**
+     * Get Amasty "Pay for everything" (items, shipping and tax) additional configuration property.
+     * True by default if configuration is not explicitly set to false.
+     * The plugin currently supports both Amasty configurations,
+     * "Use a gift card to pay for shipping" and "Use a gift card to pay for tax" set to either "Yes" or "No".
+     * Any "Yes" + "No" combination is not supported by Bolt.
+     *
+     * @return bool
+     */
+    public function getAmastyPayForEverything()
+    {
+        $amastyGiftCardConfig = $this->configHelper->getAmastyGiftCardConfig();
+        if ($amastyGiftCardConfig && @$amastyGiftCardConfig->payForEverything === false) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Get Amasty Gift Card codes, stored comma separated in total title field, return as array.
+     *
+     * @param array $totals totals array collected from quote
+     * @return array
+     */
+    public function getAmastyGiftCardCodesFromTotals($totals) {
+        $giftCardCodes = explode(',', $totals[self::AMASTY_GIFTCARD]->getTitle());
+        return array_filter(array_map('trim', $giftCardCodes));
```

Comment:
The filter doesn't seem necessary here? Do we typically get whitespace in our gift card code csv?

Reply:
@matt-thomason removed